pkg/networkservice/common/begin/client.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2021 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package begin
+
+import (
+	"context"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type beginClient struct {
+	clientMap
+}
+
+// NewClient - creates a new begin chain element
+func NewClient() networkservice.NetworkServiceClient {
+	return &beginClient{}
+}
+
+func (b *beginClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (conn *networkservice.Connection, err error) {
+	// No connection.ID, no service
+	if request.GetConnection().GetId() == "" {
+		return nil, errors.New("request.EventFactory.Id must not be zero valued")
+	}
+	// If some other EventFactory is already in the ctx... we are already running in an executor, and can just execute normally
+	if fromContext(ctx) != nil {
+		return next.Client(ctx).Request(ctx, request, opts...)
+	}
+	eventFactoryClient, _ := b.LoadOrStore(request.GetConnection().GetId(),
+		newEventFactoryClient(
+			ctx,
+			func() {
+				b.Delete(request.GetRequestConnection().GetId())
+			},
+			opts...,
+		),
+	)
+	<-eventFactoryClient.executor.AsyncExec(func() {
+		// If the eventFactory has changed, usually because the connection has been Closed and re-established
+		// go back to the beginning and try again.
+		currentConnClient, _ := b.LoadOrStore(request.GetConnection().GetId(), eventFactoryClient)
+		if currentConnClient != eventFactoryClient {
+			conn, err = b.Request(ctx, request)
```

Comment:
I feel it can be super useful if we add logging on this branch because` trace` will not cover this call. Also each return statement can be covered by logging, that can help us to determine potential bugs.